### ISSUE TITLE
Library instrumentation: Not possible to use Abstractions without implementation

### ISSUE THREAD

**Issue by lmolkova** [open-telemetry/opentelemetry-dotnet#93]
So in a typical library instrumentation scenario, the library starts/stops spans with something like (in java).

```java
span = OpenTelemetry.getTracer().spanBuilder("name").start()
```

check out https://github.com/open-telemetry/opentelemetry-java/blob/3a7e63426059f0c4202362c61e6c093851697d4a/api/src/main/java/io/opentelemetry/OpenTelemetry.java - it is a registry of default implementations with dynamic discovery 

Without implementation Span becomes `BlankSpan`, with implementation referenced, Span is a real one.

in dotnet  we have `Tracing` and it is part of `OpenTelemetry` package. I.e. library instrumentation is not possible without taking dependency on the implementation.

**Comment by lmolkova** [comment]
I've been playing with it.  Here is the raw proposal on how to fix it using delegates

```csharp

// api
Tracing
{
  public delegate ITracer GetTracerCallback();
  public static GetTracerCallback RegisterTracer;   

  public static ITracer Tracer
  {
     get
     {
       if (tracer == null)
          tracer = RegisterTracer != null ? RegisterTracer.Invoke() : new NoopTracer();
       return tracer;
     }
}

// sdk
Setup 
{
  Tracing.RegisterTracer = () => new Tracer();
}
```

pros: fast and easy, no reflection
cons: need to make sure SDK calls `Tracing.RegisterTracer = () => new Tracer();`

I've also played with reflection (and of course don't like it).
cons: very inefficient (can optimize it by looking for the known implementation and falling back to full type search)
also, it requires SDK dll to be loaded at the moment (i.e. client app need to call it first). considering this, there is no additional benefit in using reflection over delegates.
```csharp 
  var tracers = AppDomain.CurrentDomain
                         .GetAssemblies()
                         .SelectMany(a => a.GetExportedTypes())
                         .Where(p => typeof(ITracer).IsAssignableFrom(p) && p.IsClass && !p.IsAbstract)
                         .Where(t => t != typeof(NoopTracer)).ToArray();
if (tracers.Any())
{
  var tracerImpl = tracers[0];
  tracer = (ITracer)Activator.CreateInstance(tracerImpl);
}
```

Thoughts anyone? 
@SergeyKanzhelev @austinlparker @discostu105 @trwegner

**Comment by bruno-garcia** [comment]
Sorry to jump in uninvited, I hope it's OK I share some thoughts:
The delegate approach is less DI friendly but I guess that ship has sailed. All these static's would have to be set before registering anything in the container.
One nit is, just to be sure, although I'm sure it's just an example but there's a race condition in that the static field (would be a property?) could have been swapped out in between checking for null and invoking it.
```csharp
 => RegisterTracer?.Invoke() ?? new NoopTracer();
```
I'm also not a fan of the static field but the API seems to have been designed like that also to be somehow unified across other languages, I assume.

With regards to the reflection approach, the lookup only has to run once when doing the lookup so I believe would be OK. Like ASP.NET looks for Controllers to register in the container. Usually taking an option `Assembly` in case you have types in an asm other than the "EntryAssembly".

About the assembly holding the implementation having to be loaded by the time of the lookup: Won't the exporter need to be configured so that spans can be flushed? Meaning, isn't it necessary to bootstrap the exporter before using the OTy abstractions? There could also be an API to set the implementation explicitly in which case no locating would be needed?
Another con of reflection here is that it could end up picking up _the wrong_ `ITracer` implementation and being confusing. In the ASP.NET Core approach it'd all be hidding behind some extension method `AddExportarBla()` that would register all types and do what's needed.

**Comment by austinlparker** [comment]
FWIW, the delegate approach is pretty similar to the OpenTracing `GlobalTracer` (https://github.com/opentracing/opentracing-csharp/blob/master/src/OpenTracing/Util/GlobalTracer.cs).

**Comment by lmolkova** [comment]
the biggest problem with this approach is that library/instrumentation site can cache noop tracer before user has a chance to set up the right one: it's not uncommon to get configuration/secrets over HTTP and setup tracing tools/ everything else after that.

It seems the way to go is never return noop tracer, but rather wrap it with something that in runtime can switch from noop to registeres

**Comment by austinlparker** [comment]
I'd think that if you did have some situation where there was config information or secrets that you wanted traced, you'd need to have a tracer at that point, which either means you're instantiating a tracer at that point (thus solving the problem) or that you'd expect the global tracer to be registered, which could only happen during initialization.

Part of the problem is the noop tracer itself - if we were to have a wrapper, then that wrapper would have to buffer real spans up until the moment we decided that you were never going to register a tracer (and at that point would dump the buffer and everything after that would be noop), and at that point why have a noop tracer at all? 

To condense this a bit, I see three main options:
* Create a BufferedGlobalTracer abstraction around the NoopTracer in the API that gets used after the first call for _x_ time period (span count, seconds, idk) and if a real tracer is registered during that time span, flush the buffered spans into the real one and go from there; Otherwise, drop the spans and noop everything until a real tracer is registered.
* Don't do anything in terms of a wrapper and just tell people to make sure they initialize the global tracer before anything else in their process.
* Have the noop tracer be an opt-in test thing only, and maintain a buffer on the globaltracer for spans emitted before registration is complete.

**Comment by lmolkova** [comment]
I don't mind *not* tracing things that happen before tracer is set up. 

The major problem is that libraries with do something like `private readonly ITracer tracer = OpenTelemetry.Tracer` and then, when libraries are loaded in a specific order (sometimes and unpredictable and not controlled) it leads to Noop tracer cached for the lifetime of the library.

So assuming we don't need to buffer spans, wrapped tracer just forwards calls to noop or registered tracer (and it could be the noop tracer by default). This seems to be pretty straightforward and we'll just need to make it as efficient as we can.

**Comment by austinlparker** [comment]
That seems fine, yeah. We just need to make sure that `OpenTelemetry.GlobalTracer` or whatever goes to the right thing, like you said.

**Comment by lmolkova** [comment]
Fixed in #267
